polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -1716,6 +1717,270 @@ class RaiseSetCaptured : public OpHostRaisePattern<LLVM::VarAnnotation> {
 
   AccessorTypeTag accessorTypeTag;
 };
+
+/// Starting from a `sycl.host.handler.set_kernel` op, this pattern discovers
+/// the corresponding `sycl.host.handler.set_nd_range` and
+/// `sycl.host.set_captured` ops in the CGF. The pattern relies on variable
+/// annotations in SYCL's `handler.hpp` to find the number and kinds of kernel
+/// parameters. This info is used to verify that suitable values/entities have
+/// been captured.
+///
+/// Known limitations:
+/// - This pattern only matches `parallel_for` launches with the kernel passed
+///   as a lambda function.
+/// - It requires that the actual `parallel_for` call was inlined, so that the
+///   precursor ops are all in the same function (= CGF).
+/// - Launches without arugments are currently not raised due to their limited
+///   practical relevance.
+class RaiseScheduleKernel
+    : public OpHostRaisePattern<sycl::SYCLHostHandlerSetKernel> {
+public:
+  using OpHostRaisePattern<sycl::SYCLHostHandlerSetKernel>::OpHostRaisePattern;
+
+  LogicalResult matchAndRewrite(sycl::SYCLHostHandlerSetKernel op,
+                                PatternRewriter &rewriter) const final {
+    // `set_kernel` op can only be raised in a CGF.
+    auto enclosingFunc = op->getParentOfType<LLVM::LLVMFuncOp>();
+    assert(enclosingFunc);
+    auto handler = op.getHandler();
+    assert(handler == getHandler(enclosingFunc));
+
+    // Discover `set_nd_range` ops in the CGF. We are currently only interested
+    // in `parallel_for` launches, hence there should be exactly one
+    // `set_nd_range` op in the function.
+    auto setNDRangeOps =
+        enclosingFunc.getOps<sycl::SYCLHostHandlerSetNDRange>();
+    if (std::distance(setNDRangeOps.begin(), setNDRangeOps.end()) != 1)
+      return failure();
+    auto range = *setNDRangeOps.begin();
+
+    // Extract the number of parameters and the pointer to the
+    // `kernel_signatures` data structure from annotations in the current block.
+    auto failureOrParamInfo = getKernelParameterInfo(op->getBlock());
+    if (failed(failureOrParamInfo))
+      return failure();
+    auto [numParams, paramDesc] = *failureOrParamInfo;
+
+    // Parse `kernel_signatures` to get the parameter kinds.
+    auto failureOfParamKinds = getParameterKinds(numParams, paramDesc);
+    if (failed(failureOfParamKinds))
+      return failure();
+    auto paramKinds = *failureOfParamKinds;
+
+    // Adjust `num_params` if the last entry is a sentinel.
+    if (numParams > 0 && paramKinds.back() == kernel_param_kind_t::kind_invalid)
+      --numParams;
+
+    // No point in trying to raise a `schedule_kernel` without arguments.
+    if (numParams == 0)
+      return failure();
```

Comment:
Aren't we always expecting at least one argument and the last one to always be a sentinel?

Reply:
Yes, I just dropped this. I thought that I'd seen the integration header's number of parameter include the sentinel in some situations, but couldn't reproduce anymore.